confidential-identity/src/sign.rs

```diff
@@ -0,0 +1,149 @@
+//! A modified version of https://github.com/dalek-cryptography/ed25519-dalek which uses
+//! custom base point instead of a default base point.
+
+use crate::errors::{ErrorKind, Fallible};
+use curve25519_dalek::{
+    ristretto::{CompressedRistretto, RistrettoPoint},
+    scalar::Scalar,
+};
+use sha3::{digest::FixedOutput, Digest, Sha3_512};
+use zeroize::Zeroize;
+
+/// A Schnorr secret key.
+///
+/// Instances of this secret are automatically overwritten with zeroes when they
+/// fall out of scope.
+#[derive(Zeroize)]
+#[zeroize(drop)] // Overwrite secret key material with null bytes when it goes out of scope.
+pub struct SecretKey {
+    pub(crate) key: Scalar,
+    pub(crate) nonce: [u8; 32],
+}
+
+impl SecretKey {
+    /// In the original implementation: https://docs.rs/ed25519-dalek/1.0.1/src/ed25519_dalek/secret.rs.html#200
+    /// The nonce is obtained by hashing the secret key and using the upper 32 bits of the result.
+    /// The secret key is set as lower half of the result with some bit manipulation.
+    ///
+    /// Here, we do the same for the nonce, but use the secret key as is.
+    pub fn new(key: Scalar) -> Self {
+        let nonce_from_secret = Sha3_512::default().chain(&key.as_bytes()).fixed_result();
+
+        let mut nonce = [0u8; 32];
+        nonce[..].copy_from_slice(&nonce_from_secret[32..]);
+
+        Self { key, nonce }
+    }
+}
+
+/// A Schnorr public key.
+#[derive(Copy, Clone, Default, Eq, PartialEq)]
+pub struct PublicKey {
+    pub(crate) key: RistrettoPoint,
+}
+
+/// Stores the Schnorr signature for verifying the wellformedness of scope_id.
+#[derive(Debug)]
+#[allow(non_snake_case)]
+pub struct Signature {
+    pub(crate) R: CompressedRistretto,
+    pub(crate) s: Scalar,
+}
+
+impl SecretKey {
+    /// Perform a schnorr signature using a custom base point.
+    ///
+    /// # Return
+    ///
+    /// Returns the signature.
+    #[allow(non_snake_case)]
+    pub fn sign(
+        &self,
+        message: &[u8],
+        public_key: &PublicKey,
+        base_point: &RistrettoPoint,
+    ) -> Signature {
+        let mut h = Sha3_512::new();
+        let R: CompressedRistretto;
+        let r: Scalar;
+        let s: Scalar;
+        let k: Scalar;
+
+        h.input(&self.nonce);
+        h.input(&message);
+
+        r = Scalar::from_hash(h);
+        R = (&r * base_point).compress();
```

Comment:
Does this mean all signatures by the given secret will share the same `r`? 
"....Sharing a nonce is a deadly blow to Schnorr signatures"

Reply:
Yeah, I actually implemented another version that uses a randomized signature in https://github.com/PolymathNetwork/cryptography/pull/121/commits/67f9cb581d188816277288c0b212396c42a33f61 but I reverted it.
I agree with your comment, but my question is why are the current 3rd party libraries secure? Because if we were to use those libraries our nonces would be deterministic as well.
In fact, we did do this in the previous implementation of our CIL: https://github.com/PolymathNetwork/cryptography/blob/93c8451516e1d9ae342b8c7296056460aa2e425f/confidential-identity/src/claim_proofs.rs#L185

Reply:
In ed25519_dalek, the nonce is part of the secret key struct, which means when it gets generated it remains the same for all signatures: https://github.com/dalek-cryptography/ed25519-dalek/blob/925eb9ea56192053c9eb93b9d30d1b9419eee128/src/secret.rs#L240 They even have a function for generating it from the secret key deterministically: https://github.com/dalek-cryptography/ed25519-dalek/blob/925eb9ea56192053c9eb93b9d30d1b9419eee128/src/secret.rs#L265

Reply:
Similarly, Schnorrkel has a function for generating nonce from the secret key deterministically https://github.com/w3f/schnorrkel/blob/e248ee2e7f1d96a8edfe4d5455c6d23782b9923d/src/keys.rs#L145

Reply:
Because of these reasons, I reverted my changes and made the nonce generation deterministic.

Reply:
Oh, and one final thing to add: In all these implementations, the none is random and is NOT shared. It is only generated deterministically and remains the same for multiple signatures with the same secret key.

Reply:
From offline conversations:

The source of the confusion is that we are talking about two different nonces. The nonce which should change for every message is the value `k` in the implementation and it is change with every message. The nonce that is stored alongside the secret key is the seed for the deterministic randomness. Here is how it works.

When signing a `message` with a `secret key` the following are done:

1. `r = hash(nonce || message)` This ensures that different messages create different random values even with a fixed initial seed, `nonce`
2. `R = H(A) * r` where `H(A)` is the base_point in the implementation.
3. `k = hash(R || public_key || message)` Again, k will change if message has changed.
4. `s = k * secret_key + r`
5. The signature is (R, s)

Note that 
1. same message + same secret key -> same signature
2. same message + diff secret key -> diff signature: (r, R, k, s) will be different since changing secret key changes the nonce which changes r which changes everything else.
3. diff message + same secrete key -> diff signature: (r, R, k, s) will different since changing message changes r which changes everything else.

------------

Verification:
1. `k = hash(R, public_key || message)`
2. `R_verif = k * -public_key  + s * H(A)`
3. Check if `R_verif == R`

expanding `R_verif = -k * H(A) * secret_key + k * secret_key * H(A) + r*H(A) = r * H(A) = R`